File: app/src/main/java/com/example/music/ui/library/LibraryFragment.java
package com.example.music.ui.library;

import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import androidx.annotation.NonNull;
import androidx.fragment.app.Fragment;
import androidx.lifecycle.ViewModelProvider;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import com.example.music.MainViewModel;
import com.example.music.R;
import com.example.music.data.LocalMusicLoader;
import com.example.music.data.model.Track;
import com.example.music.ui.adapters.LibraryAdapter;
import com.example.music.ui.home.HomeViewModel;
import android.Manifest;
import android.content.pm.PackageManager;
import android.os.Build;
import androidx.activity.result.ActivityResultLauncher;
import androidx.activity.result.contract.ActivityResultContracts;
import androidx.core.content.ContextCompat;
import java.util.List;

public class LibraryFragment extends Fragment {

    private LibraryAdapter adapter;
    private HomeViewModel homeViewModel;
    private MainViewModel mainViewModel;

    private final ActivityResultLauncher<String> requestPermissionLauncher = registerForActivityResult(
            new ActivityResultContracts.RequestPermission(), isGranted -> {
                if (isGranted) {
                    loadLocalMusic();
                }
            });

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        return inflater.inflate(R.layout.fragment_library, container, false);
    }

    @Override
    public void onViewCreated(@NonNull View view, @androidx.annotation.Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

        homeViewModel = new ViewModelProvider(this).get(HomeViewModel.class);
        mainViewModel = new ViewModelProvider(requireActivity()).get(MainViewModel.class);

        RecyclerView recyclerView = view.findViewById(R.id.recycler_library);
        recyclerView.setLayoutManager(new LinearLayoutManager(getContext()));
        adapter = new LibraryAdapter();
        recyclerView.setAdapter(adapter);

        adapter.setOnItemClickListener(track -> {
            mainViewModel.playTrack(track);
        });

        view.findViewById(R.id.btn_cloud).setOnClickListener(v -> {
            mainViewModel.loadCloudTracks();
        });

        mainViewModel.getCloudTracks().observe(getViewLifecycleOwner(), tracks -> {
            if (tracks != null && !tracks.isEmpty()) {
                adapter.setItems(tracks);
                mainViewModel.setPlaylist(tracks);
            }
        });

        checkPermissionsAndLoad();
    }

    private void checkPermissionsAndLoad() {
        String permission = Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU ? Manifest.permission.READ_MEDIA_AUDIO
                : Manifest.permission.READ_EXTERNAL_STORAGE;

        if (ContextCompat.checkSelfPermission(requireContext(), permission) == PackageManager.PERMISSION_GRANTED) {
            loadLocalMusic();
        } else {
            requestPermissionLauncher.launch(permission);
        }
    }

    private void loadLocalMusic() {
        List<Track> localTracks = LocalMusicLoader.loadLocalTracks(requireContext());
        if (!localTracks.isEmpty()) {
            adapter.setItems(localTracks);
            mainViewModel.setPlaylist(localTracks);
        } else {
            // Fallback to charts if no local music found
            homeViewModel.getTracks().observe(getViewLifecycleOwner(), tracks -> {
                adapter.setItems(tracks);
            });
            homeViewModel.fetchCharts();
        }
    }
}

File: app/src/main/java/com/example/music/ui/library/LibraryFragment.kt
package com.example.music.ui.library

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Toast
import androidx.fragment.app.Fragment
import androidx.lifecycle.ViewModelProvider
import androidx.recyclerview.widget.LinearLayoutManager
import com.example.music.MainViewModel
import com.example.music.databinding.FragmentLibraryBinding
import com.example.music.ui.adapters.LibraryTrackAdapter

class LibraryFragment : Fragment() {

    private var _binding: FragmentLibraryBinding? = null
    private val binding get() = _binding!!

    private lateinit var mainViewModel: MainViewModel
    private lateinit var libraryAdapter: LibraryTrackAdapter

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        _binding = FragmentLibraryBinding.inflate(inflater, container, false)
        return binding.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        mainViewModel = ViewModelProvider(requireActivity())[MainViewModel::class.java]

        setupRecyclerView()
        setupButtons()
        observeData()

        // Auto-load saved tracks on open
        mainViewModel.refreshCloudLibrary()
    }

    private fun setupRecyclerView() {
        libraryAdapter = LibraryTrackAdapter()
        binding.recyclerLibrary.layoutManager = LinearLayoutManager(context)
        binding.recyclerLibrary.adapter = libraryAdapter

        libraryAdapter.setOnItemClickListener { track ->
            mainViewModel.setPlaylist(mainViewModel.savedTracks.value ?: emptyList())
            mainViewModel.playTrack(track)
        }
    }

    private fun setupButtons() {
        binding.btnConnectTest.setOnClickListener {
            mainViewModel.testConnection()
        }

        binding.btnSaveToSupabase.setOnClickListener {
            val current = mainViewModel.currentTrack.value
            if (current != null) {
                mainViewModel.saveCurrentTrackToSupabase()
            } else {
                Toast.makeText(context, "Play a track first to save it ☁️", Toast.LENGTH_SHORT).show()
            }
        }

        binding.btnLoadSaved.setOnClickListener {
            mainViewModel.refreshCloudLibrary()
        }
    }

    private fun observeData() {
        mainViewModel.savedTracks.observe(viewLifecycleOwner) { tracks ->
            libraryAdapter.setTracks(tracks)
        }

        mainViewModel.savingInProgress.observe(viewLifecycleOwner) { loading ->
            binding.progressSupabase.visibility = if (loading) View.VISIBLE else View.GONE
        }

        mainViewModel.supabaseMessage.observe(viewLifecycleOwner) { message ->
            if (!message.isNullOrEmpty()) {
                Toast.makeText(context, message, Toast.LENGTH_SHORT).show()
                mainViewModel.clearSupabaseMessage()
            }
        }
    }

    override fun onDestroyView() {
        super.onDestroyView()
        _binding = null
    }
}
